### ISSUE TITLE
Remove unnecessary byte-to-String conversion for JSON parsing

### ISSUE THREAD

**Issue by chanseokoh** [GoogleContainerTools/jib#812]
For example, in `RegistryAuthenticator` we have

```java
String responseString = Blobs.writeToString(response.getBody());
...
JsonTemplateMapper.readJson(responseString, AuthenticationResponseTemplate.class);
```
I think it can be
```java
byte[] responseBytes = Blobs.writeToByteArray(response.getBody());
...
JsonTemplateMapper.readJson(responseBytes, AuthenticationResponseTemplate.class);
```
after adding the byte-version addition to the `String`-version:
```java
  public static <...> T readJson(String jsonString, ...) ... {
    return objectMapper.readValue(jsonString, templateClass);
  }

  // Add this
  public static <...> T readJson(byte[] jsonBytes, ...) ... {
    return objectMapper.readValue(jsonBytes, templateClass);
  }
```
Of course, it's important to test that this works. But I think it will just work.

**Comment by coollog** [comment]
Hmm, I'm not sure if `ObjectMapper#readValue(byte[]...` is intended to be used. For the bytes to be turned into a JSON string that is parse-able, it would need some kind of character encoding, which isn't provided if just using raw bytes.

**Comment by chanseokoh** [comment]
Hmm... makes sense.

**Comment by briandealwis** [comment]
Conformant JSON can be encoded as UTF-8, UTF-16, or UTF-32 ([RFC7159](https://tools.ietf.org/html/rfc7159#page-9)).  `Blobs.writeToString()` assumes UTF-8. `ObjectMapper#readValue()` also takes `URL` and `File` instances, that also do not carry any explicit character set encoding.  Funnily enough though, the `JsonFactory#createParser()` methods for `URL` and `File` explicitly say they auto-detect the content encoding, whereas the `createParser(byte[])` does not.  So I think Jackson is doing the right thing.

**Comment by elharo** [comment]
JSON encoding is an absolute mess in practice. There are at least 7, possibly 8, different specifications  that cover this, and they do not agree with each other. Furthermore tools go their own way, regardless of specs.

tldr:

1. Inside Java code always pass strings to methods that expect to receive JSON, never bytes.
2. Convert bytes to strings before passing them to any JSON consuming method using an explicit encoding, never the platform default.
3. If you have a choice about the encoding to store text in, pick UTF-8.

**Comment by chanseokoh** [comment]
Alright, since our `Blob.writeToString()` always uses UTF-8 internally, I think our current code is good enough.

**Comment by elharo** [comment]
see http://seriot.ch/parsing_json.php
